Cleanup Normalizer Converter Code

ahmedkrmn opened the pull request:
Closes #126 

This PR cleans up the implementation of the [sklearn.preprocessing.Normalizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Normalizer.html) as well as its test file from the [cleanup/normalizer](https://github.com/microsoft/hummingbird/tree/cleanup/normalizer) branch.

ahmedkrmn committed:
Cleanup Normalizer Converter Code

ahmedkrmn commented:
Hi @ksaur,
Do you think I should add relative and absolute tolerance in the test?
Also is there a way to automatically generate an HTML file for the normalizer converter similar to [gbdt.html](https://github.com/microsoft/hummingbird/blob/master/doc/html/hummingbird/ml/operator_converters/gbdt.html) or is it written manually?

ksaur commented:
Thanks for your PR!

No worries on the HTML doc for now; we are working on automating that process and it will be generated at the next release


For`tests/test_sklearn_normalizer_converter.py ` line 25  you should have:

```
np.testing.assert_allclose(model.transform(data), pytorch_model(torch.from_numpy(data)).data.numpy(), rtol=1e-06, atol=1e-06)
```
 
But…when I tried this with your code (to make sure I was telling you correctly), I saw that I got an infinite loop/`StopIteration` inside `_container.py`.  This led to us doing a quick bugfix just now.   (We hadn't hit this bug before because normalizer has no param, while the other models have the param)

So hopefully if you merge the changes from PR #146, it should "just work".

ksaur commented on hummingbird/ml/operator_converters/normalizer.py:
```diff
@@ -0,0 +1,31 @@
+        elif self.norm == "max":
+            return x / torch.max(torch.abs(x), dim=1, keepdim=True)[0]
+        else:
+            raise RuntimeError("Unsupported norm: {0}".format(self.norm))
```
Please also write a test for `raise RuntimeError("Unsupported norm: {0}".format(self.norm))`.
you can see some example of error test cases [here](https://github.com/microsoft/hummingbird/blob/master/tests/test_sklearn_gbdt_converters.py#L149).

ahmedkrmn commented:
> np.testing.assert_allclose(model.transform(data), pytorch_model(torch.from_numpy(data)).data.numpy(), rtol=1e-06, atol=1e-06)
> ```
>  
Do you think it's necessary to convert the tensor data to a numpy array? `assert_allclose` works with array_like data structures so they work with tensors out of the box.

interesaaat commented:
> > np.testing.assert_allclose(model.transform(data), pytorch_model(torch.from_numpy(data)).data.numpy(), rtol=1e-06, atol=1e-06)
> > ```
> >  
> > ```
> 
> Do you think it's necessary to convert the tensor data to a numpy array? `assert_allclose` works with array_like data structures so they work with tensors out of the box.

Yes no need to do it. On the result instead you should do `.cpu().numpy()`.

ksaur commented:
Sorry I pasted the old code; I meant
```
np.testing.assert_allclose(model.transform(data), pytorch_model(data).cpu().numpy(), rtol=1e-06, atol=1e-06)

```

ahmedkrmn committed:
Add test for RuntimeError

Modify the two tests to use randomly generated arrays.

ahmedkrmn commented:
Hi Karla,
I've added a test for RuntimeError as requested. I've also changed the `data` array in `test_normalizer_converter` to be randomly generated.
I've question, I've noticed that the other tests have this line in functions that raise errors:
```python
warnings.filterwarnings("ignore")
```
What is this used for, and do we need to add it in out case?

interesaaat committed:
Update supported.py

Small change to documentation

ksaur commented:
`warnings.filterwarnings("ignore")` was for when we were getting some warnings from `scikit-learn`, but I believe that was when we were working with `0.21.3`.  (We've since moved to SKL `0.22.1` and I *think* the warnings went away but I'd have to double check.)  If you're not seeing third-party warnings, then it's fine.

ksaur reviewed (approved):
LGTM
